### PR TITLE
[Docs] Update site2/docs/functions-overview.md

### PR DESCRIPTION
**Pull request by Jennifer88huang**
### Motivation
Improve doc on Pulsar Functions.

### Modifications
1. Add an example of Pulsar Function for Go.
2. Add Pulsar Functions SDK: Go.
3. Add an example of Go function logging.
4. Refine a section title from "Parallelism" as "Run instances in parallel"
5. Refine some typos.

### Documentation

  - Does this pull request introduce a new feature? (yes )
  - If yes, how is the feature documented? ( docs)

### CODE REVIEW

**Comment by Jennifer88huang** [comment]
@wolfstudy Could you please help review the content? Thank you.

File: site2/docs/functions-overview.md [hunk 1]
```diff
@@ -154,29 +172,29 @@ $ bin/pulsar-admin functions localrun \
 
 ## Fully Qualified Function Name (FQFN)
 
-Each Pulsar Function has a **Fully Qualified Function Name** (FQFN) that consists of three elements: the function's tenant, namespace, and function name. FQFN's look like this:
+Each Pulsar Functions has a **Fully Qualified Function Name** (FQFN) that consists of three elements: the function tenant, namespace, and function name. FQFN looks like this:
```
Comment (jerrypeng) [inline site2/docs/functions-overview.md]: why do we need the change "Function" to be plural? That doesn't seem correct

File: site2/docs/functions-overview.md [hunk 2]
```diff
@@ -154,29 +172,29 @@ $ bin/pulsar-admin functions localrun \
...
 tenant/namespace/name
 ```
 
-FQFNs enable you to, for example, create multiple functions with the same name provided that they're in different namespaces.
+FQFNs enable you to create multiple functions with the same name provided that they are in different namespaces.
 
 ## Configuration
 
-Pulsar Functions can be configured in two ways:
+You can configure Pulsar Functions in the following ways:
```
Comment (jerrypeng) [inline site2/docs/functions-overview.md]:
```suggestion
You can configure a Pulsar Function in the following ways:
```

File: site2/docs/functions-overview.md [hunk 3]
```diff
@@ -189,22 +207,22 @@ inputs:
 output: persistent://public/default/test_result
 ```
 
-You can also mix and match configuration methods by specifying some function attributes via the CLI and others via YAML configuration.
+You can specify some function attributes with CLI, and specify other attributes in YAML configuration.
```
Comment (jerrypeng) [inline site2/docs/functions-overview.md]:
```suggestion
You can specify function attributes via CLI arguments or in a configuration file in YAML format.
```

File: site2/docs/functions-overview.md [hunk 4]
```diff
@@ -189,22 +207,22 @@ inputs:
...
-The Pulsar Functions API enables you to create processing logic that is:
+Pulsar Functions API enables you to create processing logic that is:
 
 * Type safe. Pulsar Functions can process raw bytes or more complex, application-specific types.
 * Based on SerDe (**Ser**ialization/**De**serialization). A variety of types are supported "out of the box" but you can also create your own custom SerDe logic.
 
 ### Function context
 
-Each Pulsar Function created using the [Pulsar Functions SDK](#the-pulsar-functions-sdk) has access to a context object that both provides:
+Each Pulsar Functions created using [Pulsar Functions SDK](#the-pulsar-functions-sdk) has access to a context object that both provides:
```
Comment (jerrypeng) [inline site2/docs/functions-overview.md]:
```suggestion
Each Pulsar Function created using [Pulsar Functions SDK](#the-pulsar-functions-sdk) has access to a context object that both provides:
```

File: site2/docs/functions-overview.md [hunk 5]
```diff
@@ -216,17 +234,17 @@ Each Pulsar Function created using the [Pulsar Functions SDK](#the-pulsar-functi
 
 ### Language-native functions
 
-Both Java and Python support writing "native" functions, i.e. Pulsar Functions with no dependencies.
+Both Java and Python support writing "native" functions, which mean Pulsar Functions with no dependencies.
```
Comment (jerrypeng) [inline site2/docs/functions-overview.md]:
```suggestion
"Native"  functions are supported in Java and Python,  which means a Pulsar Function can have no dependencies.
```

File: site2/docs/functions-overview.md [hunk 6]
```diff
@@ -216,17 +234,17 @@ Each Pulsar Function created using the [Pulsar Functions SDK](#the-pulsar-functi
...
+The benefit of native functions is that they do not have any dependencies beyond what's already available in Java/Python "out of the box." The downside is that they do not provide access to the function [context](#function-context), which is necessary for a variety of functionalities, including [logging](#logging), [user configuration](#user-configuration), and more.
 
-## The Pulsar Functions SDK
+## Pulsar Functions SDK
 
-If you'd like a Pulsar Function to have access to a [context object](#function-context), you can use the **Pulsar Functions SDK**, available for both [Java](functions-api.md#functions-for-java) and [Python](functions-api.md#functions-for-python).
+To enable Pulsar Functions to access to a [context object](#function-context), you can use **Pulsar Functions SDK**, available for [Java](functions-api.md#functions-for-java), [Python](functions-api.md#functions-for-python), and [Go](functions-api.md#functions-for-go).
```
Comment (jerrypeng) [inline site2/docs/functions-overview.md]: `
```suggestion
To enable a Pulsar Function to access to a [context object](#function-context), you can use **Pulsar Functions SDK**, available for [Java](functions-api.md#functions-for-java), [Python](functions-api.md#functions-for-python), and [Go](functions-api.md#functions-for-go).
```

File: site2/docs/functions-overview.md [hunk 7]
```diff
@@ -262,18 +280,46 @@ class ContextAwareFunction(Function):
...
+func main() {
+	pf.Start(contextFunc)
+}
+
+```
+
 ## Deployment
 
-The Pulsar Functions feature was built to support a variety of deployment options. At the moment, there are two ways to run Pulsar Functions:
+Pulsar Functions support a variety of deployment options. You can run Pulsar Functions in the following ways.
```
Comment (jerrypeng) [inline site2/docs/functions-overview.md]:
```suggestion
Pulsar Functions support a variety of deployment options. You can deploy a Pulsar Function in the following ways.
```

File: site2/docs/functions-overview.md [hunk 8]
```diff
@@ -262,18 +280,46 @@ class ContextAwareFunction(Function):
...
 :---------------|:-----------
-[Local run mode](#local-run-mode) | The function runs in your local environment, for example on your laptop
-[Cluster mode](#cluster-run-mode) | The function runs *inside of* your Pulsar cluster, on the same machines as your Pulsar [brokers](reference-terminology.md#broker)
+[Local run mode](#local-run-mode) | The function runs in your local environment, for example, on your laptop.
+[Cluster mode](#cluster-run-mode) | The function runs *inside of* your Pulsar cluster, on the same machines as your Pulsar [brokers](reference-terminology.md#broker).
 
 ### Local run mode
 
-If you run a Pulsar Function in **local run** mode, it will run on the machine from which the command is run (this could be your laptop, an [AWS EC2](https://aws.amazon.com/ec2/) instance, etc.). Here's an example [`localrun`](reference-pulsar-admin.md#localrun) command:
+If you run Pulsar Functions in the **local run** mode, you run it on the machine where you run commands(for example, your laptop, an [AWS EC2](https://aws.amazon.com/ec2/) instance). The following example is about the [`localrun`](reference-pulsar-admin.md#localrun) command.
```
Comment (jerrypeng) [inline site2/docs/functions-overview.md]:
```suggestion
If you run a Pulsar Function in the **local run** mode, you run it on the machine where you run commands(for example, your laptop, an [AWS EC2](https://aws.amazon.com/ec2/) instance). The following example is about the [`localrun`](reference-pulsar-admin.md#localrun) command.
```

File: site2/docs/functions-overview.md [hunk 9]
```diff
@@ -397,20 +468,21 @@ public class ConfigMapFunction implements Function<String, Void> {
 }
 ```
 
-### Triggering Pulsar Functions
+### Trigger Pulsar Functions
 
-Pulsar Functions running in [cluster mode](#cluster-run-mode) can be [triggered](functions-deploying.md#triggering-pulsar-functions) via the [command line](#command-line-interface). With triggering you can easily pass a specific value to a function and get the function's return value *without* needing to worry about creating a client, sending a message to the right input topic, etc. Triggering can be very useful for---but is by no means limited to---testing and debugging purposes.
+You can [trigger](functions-deploying.md#triggering-pulsar-functions) Pulsar Functions running in the [cluster mode](#cluster-mode) with the [command line](#command-line-interface). After triggering Pulsar Functions, you can pass a specific value to a function and get the return value *without* creating a client, sending a message to the right input topic. Triggering is useful for, but not limited to, testing and debugging purposes.
```
Comment (jerrypeng) [inline site2/docs/functions-overview.md]:
```suggestion
You can [trigger](functions-deploying.md#triggering-pulsar-functions) a Pulsar Function running in the [cluster mode](#cluster-mode) with the [command line](#command-line-interface). When triggering a Pulsar Function, you can pass a specific value to the Function and get the return value *without* creating a client. Triggering is useful for, but not limited to, testing and debugging purposes.
```

File: site2/docs/functions-overview.md [hunk 10]
```diff
@@ -8,9 +8,9 @@ sidebar_label: Overview
 
 * consume messages from one or more Pulsar topics,
 * apply a user-supplied processing logic to each message,
-* publish the results of the computation to another topic
+* publish the results of the computation to another topic.
 
-Here's an example Pulsar Function for Java (using the [native interface](functions-api.md#java-native-functions)):
+The following is an example of Pulsar Functions for Java (using the [native interface](functions-api.md#java-native-functions)).
```
Comment (jerrypeng) [inline site2/docs/functions-overview.md]:
```suggestion
The following is an example of a Pulsar Function written in Java (using the [native interface](functions-api.md#java-native-functions)).
```

File: site2/docs/functions-overview.md [hunk 11]
```diff
@@ -21,40 +21,58 @@ public class ExclamationFunction implements Function<String, String> {
 }
 ```
 
-Here's an equivalent function in Python (also using the [native interface](functions-api.md#python-native-functions)):
+The following is an example of Pulsar Functions for Python (using the [native interface](functions-api.md#python-native-functions)).
```
Comment (jerrypeng) [inline site2/docs/functions-overview.md]:
```suggestion
The following is an example of a Pulsar Function written in Python (using the [native interface](functions-api.md#python-native-functions)).
```

File: site2/docs/functions-overview.md [hunk 12]
```diff
@@ -21,40 +21,58 @@ public class ExclamationFunction implements Function<String, String> {
...
 
 ```python
 def process(input):
     return "{0}!".format(input)
 ```
+The following is an example of Pulsar Functions for Go.
```
Comment (jerrypeng) [inline site2/docs/functions-overview.md]:
```suggestion
The following is an example of a Pulsar Functions written in Go.
```

File: site2/docs/functions-overview.md [hunk 13]
```diff
@@ -21,40 +21,58 @@ public class ExclamationFunction implements Function<String, String> {
...
+	fmt.Println(string(in) + "!")
+	return nil
+}
+
+func main() {
+	pf.Start(HandleRequest)
+}
+```
+
+Functions are executed each time a message is published to the input topic. For example, if a function is listening to the `tweet-stream` topic, the function runs each time when a message is published to that topic.
```
Comment (jerrypeng) [inline site2/docs/functions-overview.md]:
```suggestion
A Pulsar Function is executed each time a message is published to it's input topic. For example, if a function has an input topic called `tweet-stream`, the function runs each time a message is published to `tweet-stream`.
```

File: site2/docs/functions-overview.md [hunk 14]
```diff
@@ -21,40 +21,58 @@ public class ExclamationFunction implements Function<String, String> {
...
 
 ## Goals
 
-The core goal behind Pulsar Functions is to enable you to easily create processing logic of any level of complexity without needing to deploy a separate neighboring system (such as [Apache Storm](http://storm.apache.org/), [Apache Heron](https://apache.github.io/incubator-heron), [Apache Flink](https://flink.apache.org/), etc.). Pulsar Functions is essentially ready-made compute infrastructure at your disposal as part of your Pulsar messaging system. This core goal is tied to a series of other goals:
+With Pulsar Functions, you can create complex processing logic without deploying a separate neighboring system (such as [Apache Storm](http://storm.apache.org/), [Apache Heron](https://apache.github.io/incubator-heron), [Apache Flink](https://flink.apache.org/)). Pulsar Functions are computing infrastructure of Pulsar messaging system. The core goal is tied to a series of other goals:
```
Comment (jerrypeng) [inline site2/docs/functions-overview.md]: Correct usage of Pulsar Functions here to reference the framework

**Comment by Jennifer88huang** [comment]
@jerrypeng Thank you very much for your comments.
1. Concerning the personal pronouns, we use second-person pronoun in technical writing. You can refer to the following content in *IBM Style Guide*. 
Use the pronouns that are appropriate for the person of the verb. In technical information, follow these general guidelines for personal pronouns:
- Avoid the first-person pronouns I and we, except in these situations:
° In the question portion of frequently asked questions (FAQs)
° In articles, white papers, or documents that have listed authors and in which the authors describe their own actions or opinions
- Use the second-person pronoun you as much as possible. The subject of an imperative sentence is understood to be you.
- Avoid third-person pronouns that are gender specific.
2. Concerning "Pulsar Functions", I agree with you. So if it refers to framework, we can use "Pulsar Functions" as plural form. If it refer to individual functions, could we use it as "a Pulsar function"(using lowercase "f")?

**Comment by jerrypeng** [comment]
@Jennifer88huang using second-person pronoun is ok with me.  I just know from writing academic papers for IEEE or ACM conferences, we always use third person. I don't think I have ever read a academic paper written in second-person.  Though our documentation doesn't need to be as formal as an academic paper.

> So if it refers to framework, we can use "Pulsar Functions" as plural form. 

Yes

> If it refer to individual functions, could we use it as "a Pulsar function"(using lowercase "f")?

Yes.  In some places we use Pulsar Function (capital case "f") in other places we use Pulsar function  (lowercase "f"). We probably should be consistent throughout our documentation, though I am not sure which is more appropriate, capital case "f" or lowercase "f".

Though for Azure Functions, its documentation uses a capital case "f":

https://azure.microsoft.com/en-us/services/functions/

Maybe we should just do that?

**Comment by Jennifer88huang** [comment]
> @jerrypeng  using second-person pronoun is ok with me. I just know from writing academic papers for IEEE or ACM conferences, we always use third person. I don't think I have ever read a academic paper written in second-person. Though our documentation doesn't need to be as formal as an academic paper.
When we write academic paper, we use third person to indicate they are objective opinions. It's requirement of academic paper. When we write technical documentation (user manuals), our target readers are our users, we need adopt user-focus style, and involve them as much as possible. That's why we use second-person pronoun in technical writing.
When you read user manuals at IEEE website, you will find that their user manuals also adopt second-person pronoun style. The following is an example from IEEE.
image
This does not mean that technical documentation does not need to be as formal as an academic paper. They just adopt different writing styles, in order to achieve different goals.
When you read Microsoft technical docs, you will find that they also adopt second-person pronoun.
image

> Yes. In some places we use Pulsar Function (capital case "f") in other places we use Pulsar function (lowercase "f"). We probably should be consistent throughout our documentation, though I am not sure which is more appropriate, capital case "f" or lowercase "f".
> 
> Though for Azure Functions, its documentation uses a capital case "f":
> 
> https://azure.microsoft.com/en-us/services/functions/
> 
Thank you very much for your information. "Function" is really a challenging noun here. And I totally agree with you that we need keep it consistent throughout our docs.

I read the Azure Functions doc you recommend, they use "Functions" and "functions" at the same time. I guess when they refer to "Azure Functions/Durable Functions/Functions" (as proper noun), they use the uppercase letter; when they refer to other functions (especially created by users), they use lowercase letter. 
image
Could we ask others to discuss on this issue and decide how to use this word in our docs? After decision, I can update it throughout.

**Comment by jerrypeng** [comment]
@Jennifer88huang thanks for information about how and when to use third and second person in technical writing.

In regards to "Functions" vs "functions", after examining Azure documentation in more detail, it uses capital "f" in "functions" when they are referring to the name of the product i.e. "Azure Functions" or "Durable Functions".  It uses a lower case "f" for "function" when referring to a general or common concept. "Serverless functions" is not a brand name, it is the term for a general concept/framework, thus I think it is spelled with a lower case "f".  

I think its reasonable for us spell Pulsar Functions with a capital "f", when we are referring to this specific product/framework.  We can use the lower case "functions", when we refer to general concept of a function e.g. math function

@merlimat @sijie @srkukarni @rdhabalia thoughts?

**Comment by sijie** [comment]
@jerrypeng +1

**Comment by Jennifer88huang** [comment]
If there is no disagreement with the usage of "Pulsar Functions", "a Pulsar function", I'll update it throughout.

**Comment by jerrypeng** [comment]
@Jennifer88huang sounds good to me!  Please also update https://github.com/apache/pulsar/pull/4529.  Thanks you!

**Comment by Jennifer88huang** [comment]
Great, sure, I'll update it soon.

**Comment by Jennifer88huang** [comment]
I've update the usage of "functions" in this file, if i have any misunderstanding or incorrect usage, feel free to correct it. The principle follows:
- Pulsar Functions / a Pulsar Function (refer to product/feature specific to Pulsar)
- function/functions (refer to common/individual functions)

**Comment by jerrypeng** [comment]
rerun integration tests

**Comment by Jennifer88huang** [comment]
@jerrypeng Thank you very much for your comments and advice~

**Comment by sijie** [comment]
run integration tests

**Comment by jerrypeng** [comment]
rerun integration tests

**Comment by Jennifer88huang** [comment]
run integration tests

**Comment by sijie** [comment]
well done @Jennifer88huang

**Pull request merged**